accord-core/src/main/java/accord/impl/SimpleProgressLog.java

```diff
@@ -0,0 +1,547 @@
+package accord.impl;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+
+import accord.api.Key;
+import accord.api.ProgressLog;
+import accord.api.Result;
+import accord.coordinate.CheckShardProgress;
+import accord.local.Command;
+import accord.local.CommandStore;
+import accord.local.Node;
+import accord.local.Node.Id;
+import accord.local.Status;
+import accord.messages.Apply;
+import accord.messages.Callback;
+import accord.messages.CheckStatus.CheckStatusOk;
+import accord.messages.CheckStatus.CheckStatusOkFull;
+import accord.messages.CheckStatus.IncludeInfo;
+import accord.messages.Reply;
+import accord.txn.Ballot;
+import accord.txn.Dependencies;
+import accord.txn.Timestamp;
+import accord.txn.Txn;
+import accord.txn.TxnId;
+import accord.txn.Writes;
+
+import static accord.coordinate.CheckHomeStatus.checkHomeStatus;
+import static accord.coordinate.CheckShardProgress.checkShardProgress;
+import static accord.coordinate.InformHomeOfTxn.inform;
+import static accord.coordinate.MaybeRecover.maybeRecover;
+import static accord.impl.SimpleProgressLog.CoordinateSendAndCheck.sendAndCheck;
+import static accord.impl.SimpleProgressLog.ExternalStatus.Committed;
+import static accord.impl.SimpleProgressLog.ExternalStatus.NonHomePreAccept;
+import static accord.impl.SimpleProgressLog.ExternalStatus.NonHomeSafe;
+import static accord.impl.SimpleProgressLog.ExternalStatus.NonHomeWaitingOn;
+import static accord.impl.SimpleProgressLog.ExternalStatus.ReadyToExecute;
+import static accord.impl.SimpleProgressLog.ExternalStatus.ExecutedOnAllShards;
+import static accord.impl.SimpleProgressLog.ExternalStatus.Uncommitted;
+import static accord.impl.SimpleProgressLog.ExternalStatus.WaitingForExecutedOnAllShards;
+import static accord.impl.SimpleProgressLog.InternalStatus.Done;
+import static accord.impl.SimpleProgressLog.InternalStatus.Investigating;
+import static accord.impl.SimpleProgressLog.InternalStatus.NoProgress;
+import static accord.impl.SimpleProgressLog.InternalStatus.NoProgressExpected;
+import static accord.impl.SimpleProgressLog.InternalStatus.Waiting;
+import static accord.local.Status.Executed;
+import static accord.local.Status.PreAccepted;
+
+public class SimpleProgressLog implements Runnable, Function<CommandStore, ProgressLog>
```

Comment:
One of my medium term goals is refactoring the synchronisation logic a little so that the command stores are always safe for thread-unsafe code, as I think most code should be thread-unsafe by default. In this case we allocate one of these per command store, so I think the only issue we could have we be caused by not guarding every execution by the synchronised keyword...

I think this would be fine for production use for in-memory-only workloads, and would be fine for reference implementations to use (and for us to use initially in C* while we develop everything else), but I agree some strong caveats should be listed.

Reply:
This data structure exists per-commandstore, and so a thread unsafe data structure should be _preferred_ IMO. The thread model needs to be revisited at some point in the near future, however, to ensure this can be done safely and cleanly, but I do not think there is anything to caveat here besides that it is non-durable (perhaps we should call it a `NonDurableProgressLog`).